### DISCUSSION TITLE
STM32 QSPI Flash memory mapped mode

### DISCUSSION THREAD

**Discussion by RIPHeisenberg** [zephyrproject-rtos/zephyr#77875] [Q&A]
Hi everyone,

I am currently trying to use the QSPI Flash API of the STM32H747I_DISCOVERY_M7 chip.

Zephyr version 3.7 LTS
Board STM32H747I_DISCOVERY_M7

What I have tried so far:
KCONFIG:
CONFIG_STDOUT_CONSOLE=y
CONFIG_FLASH=y
CONFIG_SPI=y

CONFIG_GPIO=y
CONFIG_FLASH=y
CONFIG_FLASH_STM32_QSPI=y
CONFIG_STM32_MEMMAP=y

CONFIG_PRINTK=y
CONFIG_SHELL=y
CONFIG_LOG=y
CONFIG_FLASH=y
CONFIG_FLASH_SHELL=y
CONFIG_FLASH_SHELL_TEST_COMMANDS=y

Device Tree:
		quadspi: quadspi@52005000 {
			compatible = "st,stm32-qspi";
			#address-cells = < 0x1 >;
			#size-cells = < 0x0 >;
			reg = < 0x52005000 0x34 >;
			interrupts = < 0x5c 0x0 >;
			clocks = < &rcc 0xd4 0x4000 >;
			status = "okay";
			pinctrl-0 = < &quadspi_clk_pb2 &quadspi_bk1_ncs_pg6 &quadspi_bk1_io0_pd11 &quadspi_bk1_io1_pf9 
&quadspi_bk1_io2_pf7 &quadspi_bk1_io3_pf6 &quadspi_bk2_io0_ph2 &quadspi_bk2_io1_ph3 &quadspi_bk2_io2_pg9 &quadspi_bk2_io3_pg14 >;
			pinctrl-names = "default";
			dual-flash;
			mt25ql512ab1: qspi-nor-flash-1@90000000 {
				compatible = "st,stm32-qspi-nor";
				reg = < 0x90000000 0x4000000 >;
				qspi-max-frequency = < 0x44aa200 >;
				spi-bus-width = < 0x4 >;
				status = "okay";
				partitions {
					compatible = "fixed-partitions";
					#address-cells = < 0x1 >;
					#size-cells = < 0x1 >;
					slot0_partition: partition@0 {
						reg = < 0x0 0x200000 >;
						label = "image-0";
					};
				};
			};

Example: samples/drivers/spi_flash

Following output is generated: 

```
[00:00:00.000,000] <inf> flash_stm32_qspi: Reading SFDP
[00:00:00.000,000] <inf> flash_stm32_qspi: qspi-nor-flash-1@90000000: SFDP v 1.6 AP ff with 2 PH
[00:00:00.000,000] <inf> flash_stm32_qspi: PH0: ff00 rev 1.6: 16 DW @ 30
[00:00:00.000,000] <inf> flash_stm32_qspi: Reading SFDP
[00:00:00.000,000] <inf> flash_stm32_qspi: qspi-nor-flash-1@90000000: 128 MiBy flash
[00:00:00.000,000] <inf> flash_stm32_qspi: Flash - address mode: 4B
[00:00:00.000,000] <inf> flash_stm32_qspi: Quad read mode 4 instr [0x6b] supported
[00:00:00.000,000] <inf> flash_stm32_qspi: Quad read mode 7 instr [0xeb] supported
[00:00:00.000,000] <inf> flash_stm32_qspi: Quad read mode 7 instr [0xeb] will be used
[00:00:00.000,000] <inf> flash_stm32_qspi: QE requirement mode: 0
[00:00:00.000,000] <inf> flash_stm32_qspi: Quad mode enabled
[00:00:00.000,000] <inf> flash_stm32_qspi: PH1: ff84 rev 1.0: 2 DW @ 80
[00:00:00.000,000] <inf> flash_stm32_qspi: Memory-mapped NOR quad-flash at 0x90000000 (0x8000000 bytes)
*** Booting Zephyr OS build v3.7.0-506-g832261dbe5d7 ***
uart:~$ 
qspi-nor-flash-1@90000000 SPI flash testing
==========================

Perform test on single sector
Test 1: Flash erase
Flash erase failed! -22

Test 2: Flash write
Attempting to write 4 bytes
Flash write failed! -22

Perform test on multiple consecutive sectors
Test 1: Flash erase
Flash erase failed! -22

Test 2: Flash write
Attempting to write 4 bytes at offset 0x90000000
Flash write failed! -22
```

I tried it with region offset 0x990000000 and 0x0.
neither of which worked.
SPI Sector size 4096

Is it possible to use the QSPI flash in Memory mapped mode? What am I missing? Some help is highly appreciated.

Best regards
Maikel Schuurmans Stekhoven

**Comment by erwango** [comment]
@FRASTM ^^

**Comment by mariopaja** [comment]
@RIPHeisenberg if you want to implement memory mapped mode, you can have a look at this sample https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/code_relocation_nocopy 

I have already used this sample with a nucleo_h745zi_q and an external flash memory and it worked like a charm. The only think you would keep in mind is the linker needed for the flash memory. I dont know if it is implemented for your board.

**Reply by FRASTM** [reply]
> @RIPHeisenberg if you want to implement memory mapped mode, you can have a look at this sample https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/code_relocation_nocopy
> 
> I have already used this sample with a nucleo_h745zi_q and an external flash memory and it worked like a charm. The only think you would keep in mind is the linker needed for the flash memory. I dont know if it is implemented for your board.

Correct for stm32h745 or stm32h750 disco kits. However for the stm32h747i disco, I could not get the same result  and I did not find any explanation (see https://github.com/zephyrproject-rtos/zephyr/discussions/72429 ) 
Note that XiP only performs "read" operation from the quad-NOR.

**Reply by RIPHeisenberg** [reply]
Hi @mariopaja @FRASTM 

Thanks for your responds. I have some nucleo_h745 boards at work. I'll try out your suggestion. Thanks a lot for the support. I assume you add a external flash to the board via the QSPI pinout, right? The nucleo_h745 doesn't have a internal QSPI flash as far as I know.

**Reply by RIPHeisenberg** [reply]
Do you use the cube programmer to load code to the external flash? Is there a more "Zephyr" way of doing this? 
Not that I would mind using the cube programmer

**Reply by mariopaja** [reply]
@RIPHeisenberg 

> I assume you add a external flash to the board via the QSPI pinout, right? The nucleo_h745 doesn't have a internal QSPI flash as far as I know.

exactly

> Do you use the cube programmer to load code to the external flash? Is there a more "Zephyr" way of doing this? Not that I would mind using the cube programmer

you can use  **_--runner stm32cubeprogrammer_**

Just keep in mind that you need an external loader to make it work (xxxxxx.stldr). They are platform, pinout (I might be wrong on this), and flash memory specific.

**Reply by RIPHeisenberg** [reply]
Thanks! I hoped I could avoid writing a external loader. It seems like a very cumbersome process...
But it makes sense that this has to be done for a custom Flash chip.

**Reply by mariopaja** [reply]
@RIPHeisenberg I didnt write a loader either. I was lucky enough to find the right one for my test application here https://github.com/cturvey/stm32extldr

It would be nice if there would be a more straightforward way to use the flash memories or write the loaders

**Comment by FRASTM** [comment]
Hi @RIPHeisenberg 

The stm32 qspi driver cannot write or erase in memory mapped mode. Only reading is possible (memcopy). With the spi_flash sample, setting memory mapped and aborting memory mapped mode can disturb the access to the quad-spi flash.
